File: medium/0189_rotate_array/solution.cpp
#include <vector>

class Solution {
public:
  void rotate(std::vector<int> &nums, int k) {
    const int n = nums.size();
    k = k % n;

    for (auto &[left, right] :
         std::vector<std::pair<int, int>>{{0, n - 1}, {0, k - 1}, {k, n - 1}}) {
      while (left < right)
        std::swap(nums[left++], nums[right--]);
    }
  }
};

File: medium/0189_rotate_array/solution.py
from typing import List


class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        n = len(nums)
        k = k % n

        for left, right in ((0, n - 1), (0, k - 1), (k, n - 1)):
            while left < right:
                nums[left], nums[right] = nums[right], nums[left]
                left += 1
                right -= 1
